compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractSteppingTest.kt

```diff
@@ -0,0 +1,202 @@
+import org.jetbrains.kotlin.test.ConfigurationKind
+import java.io.File
+import java.io.IOException
+import java.net.Socket
+import java.nio.file.Files
+import kotlin.properties.Delegates
+
+abstract class AbstractSteppingTest : CodegenTestCase() {
+    private companion object {
+        private const val DEBUG_ADDRESS = "127.0.0.1"
+        private const val DEBUG_PORT = 5225
+    }
+
+    override fun doMultiFileTest(wholeFile: File, files: List<TestFile>) {
+        createEnvironmentWithMockJdkAndIdeaAnnotations(ConfigurationKind.ALL, *listOfNotNull(writeJavaFiles(files)).toTypedArray())
+
+        val lineNumbers = wholeFile
+            .readLines()
+            .filter { it.matches("// \\d+.*".toRegex()) }
+            .single()
+            .drop(2)
+            .trim()
+            .split(" ")
+            .map { it.trim().toInt() }
+
+        loadMultiFiles(files)
+        val classBuilderFactory = OriginCollectingClassBuilderFactory(ClassBuilderMode.FULL)
+        val generationState = GenerationUtils.compileFiles(myFiles.psiFiles, myEnvironment, classBuilderFactory)
+        classFileFactory = generationState.factory
+
+        val tempDirForTest = Files.createTempDirectory("debuggerTest").toFile()
+
+        try {
+            val classesDir = File(tempDirForTest, "classes").apply {
+                for (classFile in classFileFactory.getClassFiles()) {
+                    File(this, classFile.relativePath).mkdirAndWriteBytes(classFile.asByteArray())
+                }
+            }
+
+            val process = startDebuggeeProcess(classesDir)
+            waitUntil { isPortOpen() }
+
+            val virtualMachine = attachDebugger()
+
+            try {
+                doTest(lineNumbers, virtualMachine, classBuilderFactory, classFileFactory, generationState)
+            } finally {
+                process.destroy()
+            }
+        } finally {
+            tempDirForTest.deleteRecursively()
+        }
+    }
+
+    protected open fun doTest(
+        lineNumbers: List<Int>,
+        virtualMachine: VirtualMachine,
+        factory: OriginCollectingClassBuilderFactory,
+        classFileFactory: ClassFileFactory,
+        state: GenerationState
+    ) {
+        val manager = virtualMachine.eventRequestManager()
+        val loggedEvents = ArrayList<Event>()
+
+        var inBox = false
+        var finish = false
+        vmLoop@ while (true) {
+            val eventSet = virtualMachine.eventQueue().remove()
+            for (event in eventSet) {
+                when (event) {
+                    is VMDeathEvent, is VMDisconnectEvent -> {
+                        finish = true
+                    }
+                    is VMStartEvent -> {
+                        val classReq = manager.createClassPrepareRequest()
+                        classFileFactory.getClassFiles().map { classReq.addClassFilter(it.internalName) }
+                        classReq.setSuspendPolicy(SUSPEND_ALL)
```

Comment:
Discussed offline: Should be faster if you use `SUSPEND_NONE` for all these requests, and use `createVMDeathRequest()` (https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/com/sun/jdi/request/EventRequestManager.html#createVMDeathRequest()) to suspend before VM dies.